SetupDataPkg/Library/Examples/ConfigKnobShimExampleLib/ConfigKnobShimLibCommon.c

```diff
@@ -0,0 +1,169 @@
+                                  not, EFI_BAD_BUFFER_SIZE will be returned.
+  @param[in] ProfileDefaultValue  The profile defined default value of this knob, to be filled in if knob is not
+                                  overridden in a cached policy or variable storage.
+  @param[in] ProfileDefaultSize
+
+  @retval EFI_INVALID_PARAMETER   Input argument is null.
+  @retval EFI_BAD_BUFFER_SIZE     Caller passed in a value of ConfigKnobDataSize that does not equal
+                                  ProfileDefaultSize.
+  @retval EFI_SUCCESS             The operation succeeds.
+
+**/
+EFI_STATUS
+EFIAPI
+GetConfigKnob (
+  IN EFI_GUID  *ConfigKnobGuid,
+  IN CHAR16    *ConfigKnobName,
+  OUT VOID     *ConfigKnobData,
+  IN UINTN     ConfigKnobDataSize,
+  IN VOID      *ProfileDefaultValue,
+  IN UINTN     ProfileDefaultSize
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       VariableSize = ConfigKnobDataSize;
+
+  if ((ConfigKnobGuid == NULL) || (ConfigKnobName == NULL) || (ConfigKnobData == NULL) ||
+      (ConfigKnobDataSize == 0) || (ProfileDefaultValue == NULL) || (ProfileDefaultSize == 0))
+  {
+    DEBUG ((DEBUG_ERROR, "%a: Invalid parameter!\n", __FUNCTION__));
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  if (ProfileDefaultSize != ConfigKnobDataSize) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: ConfigKnobDataSize %u does not equal ProfileDefaultSize %u\n",
+      __FUNCTION__,
+      ConfigKnobDataSize,
+      ProfileDefaultSize
+      ));
+    Status = EFI_BAD_BUFFER_SIZE;
+    goto Exit;
+  }
+
+  // Check if it is in variable storage
+  Status = GetConfigKnobFromVariable (ConfigKnobGuid, ConfigKnobName, ConfigKnobData, &VariableSize);
+
+  // TODO: Also check attributes here
+  if (ConfigKnobDataSize != VariableSize) {
+    // we will only accept this variable if it is the correct size
+    Status = EFI_NOT_FOUND;
+  }
+
+  if (EFI_ERROR (Status)) {
+    // we didn't find the value in variable storage, which is expected if the knob has not been overridden, or
+    // the size mismatched. In either case, we default to the profile value.
+    DEBUG ((
+      DEBUG_VERBOSE,
+      "%a: failed to find config knob %a with status %r. Expected size: %u, found size: %u."
+      " Defaulting to profile defined value.\n",
+      __FUNCTION__,
+      ConfigKnobName,
+      Status,
+      ConfigKnobDataSize,
+      VariableSize
+      ));
+
+    CopyMem (ConfigKnobData, ProfileDefaultValue, ProfileDefaultSize);
+
+    // we failed to get the variable, but successfully returned the profile value, this is a success case
+    Status = EFI_SUCCESS;
+  }
+
+Exit:
+  return Status;
+}
+
+C_ASSERT (sizeof (EFI_GUID) == sizeof (config_guid_t));
+
+// Get the raw knob value
+//
+// This function is called by the generated functions and uses the generated knob metadata to call into the generic
+// GetConfigKnob function to get the knob value.
+void *
+get_knob_value (

```

Comment:
This should likely be part of the autogenerated code in ConfigServiceGenerated.h and converted to UEFI style.